doc/man/pivxd.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.6.
.TH PIVXD "1" "December 2022" "pivxd v5.5.0.0" "User Commands"
.SH NAME
pivxd \- manual page for pivxd v5.5.0.0
.SH SYNOPSIS
.B pivxd
[\fI\,options\/\fR]                     \fI\,Start PIVX Core Daemon\/\fR
.SH DESCRIPTION
PIVX Core Daemon version v5.5.0.0
.SH OPTIONS
.HP
\-?
.IP
This help message
.HP
\fB\-version\fR
.IP
Print version and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when a relevant alert is received or we see a really
long fork (%s in cmd is replaced by message)
.HP
\fB\-blocksdir=\fR<dir>
.IP
Specify directory to hold blocks subdirectory for *.dat files (default:
<datadir>)
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-checkblocks=\fR<n>
.IP
How many blocks to check at startup (default: 6, 0 = all)
.HP
\fB\-checklevel=\fR<n>
.IP
How thorough the block verification of \fB\-checkblocks\fR is (0\-4, default: 3)
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: pivx.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-paramsdir=\fR<dir>
.IP
Specify zk params directory (default: /home/fuzzbawls/.pivx\-params)
.HP
\fB\-debuglogfile=\fR<file>
.IP
Specify location of debug log file: this can be an absolute path or a
path relative to the data directory (default: debug.log)
.HP
\fB\-disablesystemnotifications\fR
.IP
Disable OS notifications for incoming transactions (default: 0)
.HP
\fB\-dbcache=\fR<n>
.IP
Set database cache size in megabytes (4 to 16384, default: 300)
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external blk000??.dat file on startup
.HP
\fB\-maxreorg=\fR<n>
.IP
Set the Maximum reorg depth (default: 100)
.HP
\fB\-maxorphantx=\fR<n>
.IP
Keep at most <n> unconnectable transactions in memory (default: 25)
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the transaction memory pool below <n> megabytes (default: 300)
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
72)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-par=\fR<n>
.IP
Set the number of script verification threads (\fB\-16\fR to 16, 0 = auto, <0 =
leave that many cores free, default: 0)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file (default: pivx.pid)
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks
.HP
\fB\-reindex\fR
.IP
Rebuild block chain index from current blk000??.dat files on startup
.HP
\fB\-resync\fR
.IP
Delete blockchain folders and resync from scratch on startup
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 1)
.HP
\fB\-forcestart\fR
.IP
Attempt to force blockchain corruption recovery on startup
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open
.HP
\fB\-asmap=\fR<file>
.IP
Specify asn mapping used for bucketing of the peers (default:
ip_asn.map). Relative paths will be prefixed by the net\-specific datadir
location.
.HP
\fB\-banscore=\fR<n>
.IP
Threshold for disconnecting misbehaving peers (default: 100)
.HP
\fB\-bantime=\fR<n>
.IP
Number of seconds to keep misbehaving peers from reconnecting (default:
86400)
.HP
\fB\-bind=\fR<addr>
.IP
Bind to given address and always listen on it. Use [host]:port notation
for IPv6
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node(s); \fB\-noconnect\fR or \fB\-connect\fR=\fI\,0\/\fR alone to
disable automatic connections
.HP
\fB\-discover\fR
.IP
Discover own IP address (default: 1 when listening and no \fB\-externalip\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR/\-noconnect)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 0)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or
\fB\-connect\fR/\-noconnect)
.HP
\fB\-listenonion\fR
.IP
Automatically create Tor hidden service (default: 1)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (default: 125)
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer, <n>*1000 bytes (default: 5000)
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer, <n>*1000 bytes (default: 1000)
.HP
\fB\-onion=\fR<ip:port>
.IP
Use separate SOCKS5 proxy to reach peers via Tor hidden services
(default: \fB\-proxy\fR)
.HP
\fB\-onlynet=\fR<net>
.IP
Only connect to nodes in network <net> (ipv4, ipv6 or onion)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
1)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port> (default: 51472 or testnet: 51474)
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. This enables Tor
stream isolation (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect
.HP
\fB\-timeout=\fR<n>
.IP
Specify connection timeout in milliseconds (minimum: 1, default: 5000)
.HP
\fB\-torcontrol=\fR<ip>:<port>
.IP
Tor control port to use if onion listening enabled (default:
127.0.0.1:9051)
.HP
\fB\-torpassword=\fR<pass>
.IP
Tor control port password (default: empty)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 0)
.HP
\fB\-natpmp\fR
.IP
Use NAT\-PMP to map the listening port (default: 0)
.HP
\fB\-whitebind=\fR<addr>
.IP
Bind to given address and whitelist peers connecting to it. Use
[host]:port notation for IPv6
.HP
\fB\-whitelist=\fR<netmask>
.IP
Whitelist peers connecting from the given netmask or IP address. Can be
specified multiple times. Whitelisted peers cannot be DoS banned and
their transactions are always relayed, even if they are already in the
mempool, useful e.g. for a gateway
.PP
Wallet options:
.HP
\fB\-createwalletbackups=\fR<n>
.IP
Number of automatic wallet backups (default: 10)
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls (default: 0)
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 100)
.HP
\fB\-legacywallet\fR
.IP
On first run, create a legacy wallet instead of a HD wallet
.HP
\fB\-maxtxfee=\fR<amt>
.IP
Maximum total fees to use in a single wallet transaction, setting too
low may abort large transactions (default: 1.00)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fees (in PIV/Kb) smaller than this are considered zero fee for
transaction creation (default: 0.0001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee (in PIV/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-rescan\fR
.IP
Rescan the block chain for missing wallet transactions on startup
.HP
\fB\-salvagewallet\fR
.IP
Attempt to recover private keys from a corrupt wallet file on startup
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-txconfirmtarget=\fR<n>
.IP
If paytxfee is not set, include enough fee so transactions begin
confirmation on average within n blocks (default: 1)
.HP
\fB\-upgradewallet\fR
.IP
Upgrade wallet to latest format on startup
.HP
\fB\-wallet=\fR<path>
.IP
Specify wallet database path. Can be specified multiple times to load
multiple wallets. Path is interpreted relative to <walletdir> if it is
not absolute, and will be created if it does not exist (as a directory
containing a wallet.dat file and log files). For backwards compatibility
this will also accept names of existing data files in <walletdir>.)
.HP
\fB\-walletdir=\fR<dir>
.IP
Specify directory to hold wallets (default: <datadir>/wallets if it
exists, otherwise <datadir>)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes (%s in cmd is replaced
by TxID)
.HP
\fB\-zapwallettxes=\fR<mode>
.IP
Delete all wallet transactions and only recover those parts of the
blockchain through \fB\-rescan\fR on startup(1 = keep tx meta data e.g. payment
request information, 2 = drop tx meta data)
.PP
Mining/Staking options:
.HP
\fB\-coldstaking=\fR<n>
.IP
Enable cold staking functionality (0\-1, default: 1). Disabled if
staking=0
.HP
\fB\-gen\fR
.IP
Generate coins (default: 0)
.HP
\fB\-genproclimit=\fR<n>
.IP
Set the number of threads for coin generation if enabled (\fB\-1\fR = all
cores, default: 1)
.HP
\fB\-minstakesplit=\fR<amt>
.IP
Minimum positive amount (in PIV) allowed by GUI and RPC for the stake
split threshold (default: 100.00)
.HP
\fB\-staking=\fR<n>
.IP
Enable staking functionality (0\-1, default: 1)
.PP
ZeroMQ notification options:
.HP
\fB\-zmqpubhashblock=\fR<address>
.IP
Enable publish hash block in <address>
.HP
\fB\-zmqpubhashtx=\fR<address>
.IP
Enable publish hash transaction in <address>
.HP
\fB\-zmqpubrawblock=\fR<address>
.IP
Enable publish raw block in <address>
.HP
\fB\-zmqpubrawtx=\fR<address>
.IP
Enable publish raw transaction in <address>
.PP
Debugging/Testing options:
.HP
\fB\-uacomment=\fR<cmt>
.IP
Append comment to the user agent string
.HP
\fB\-debug=\fR<category>
.IP
Output debugging information (default: 0, supplying <category> is
optional). If <category> is not supplied, output all debugging
information. <category> can be: net, tor, mempool, http, bench, zmq, db,
rpc, estimatefee, addrman, reindex, proxy, mempoolrej, libevent, coindb,
qt, leveldb, staking, masternode, mnbudget, zero, mnping, sapling,
sporks, validation, llmq, net_mn, dkg.
.HP
\fB\-debugexclude=\fR<category>
.IP
Exclude debugging information for a category. Can be used in conjunction
with \fB\-debug\fR=\fI\,1\/\fR to output debug logs for all categories except one or more
specified categories.
.HP
\fB\-help\-debug\fR
.IP
Show all debugging options (usage: \fB\-\-help\fR \fB\-help\-debug\fR)
.HP
\fB\-logips\fR
.IP
Include IP addresses in debug output (default: 0)
.HP
\fB\-logtimestamps\fR
.IP
Prepend debug output with timestamp (default: 1)
.HP
\fB\-logtimemicros\fR
.IP
Add microsecond precision to debug timestamps (default: 0)
.HP
\fB\-maxtipage=\fR<n>
.IP
Maximum tip age in seconds to consider node in initial block download
(default: 86400)
.HP
\fB\-minrelaytxfee=\fR<amt>
.IP
Fees (in PIV/Kb) smaller than this are considered zero fee for relaying,
mining and transaction creation (default: 0.0001)
.HP
\fB\-printtoconsole\fR
.IP
Send trace/debug info to console instead of debug.log file (default: 0)
.HP
\fB\-shrinkdebugfile\fR
.IP
Shrink debug.log file on client startup (default: 1 when no \fB\-debug\fR)
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.PP
Masternode options:
.HP
\fB\-masternode=\fR<n>
.IP
Enable the client to act as a masternode (0\-1, default: 0)
.HP
\fB\-mnconf=\fR<file>
.IP
Specify masternode configuration file (default: masternode.conf)
.HP
\fB\-mnconflock=\fR<n>
.IP
Lock masternodes from masternode configuration file (default: 1)
.HP
\fB\-masternodeprivkey=\fR<n>
.IP
Set the masternode private key
.HP
\fB\-masternodeaddr=\fR<n>
.IP
Set external address:port to get to this masternode (example:
128.127.106.235:51472). Only for Legacy Masternodes
.HP
\fB\-budgetvotemode=\fR<mode>
.IP
Change automatic finalized budget voting behavior. mode=auto: Vote for
only exact finalized budget match to my generated budget. (string,
default: auto)
.HP
\fB\-mnoperatorprivatekey=\fR<bech32>
.IP
Set the masternode operator private key. Only valid with \fB\-masternode\fR=\fI\,1\/\fR.
When set, the masternode acts as a deterministic masternode.
.PP
Node relay options:
.HP
\fB\-datacarrier\fR
.IP
Relay and mine data carrier transactions (default: 1)
.HP
\fB\-datacarriersize\fR
.IP
Maximum size of data in data carrier transactions we relay and mine
(default: 83)
.PP
Block creation options:
.HP
\fB\-blockmaxsize=\fR<n>
.IP
Set maximum block size in bytes (default: 750000)
.PP
RPC server options:
.HP
\fB\-server\fR
.IP
Accept command line and JSON\-RPC commands
.HP
\fB\-rest\fR
.IP
Accept public REST requests (default: 0)
.HP
\fB\-rpcbind=\fR<addr>
.IP
Bind to given address to listen for JSON\-RPC connections. Do not expose
the RPC server to untrusted networks such as the public internet! This
option is ignored unless \fB\-rpcallowip\fR is also passed. Port is optional
and overrides \fB\-rpcport\fR. Use [host]:port notation for IPv6. This option
can be specified multiple times (default: 127.0.0.1 and ::1 i.e.,
localhost)
.HP
\fB\-rpccookiefile=\fR<loc>
.IP
Location of the auth cookie (default: data dir)
.HP
\fB\-rpcuser=\fR<user>
.IP
Username for JSON\-RPC connections
.HP
\fB\-rpcpassword=\fR<pw>
.IP
Password for JSON\-RPC connections
.HP
\fB\-rpcauth=\fR<userpw>
.IP
Username and hashed password for JSON\-RPC connections. The field
<userpw> comes in the format: <USERNAME>:<SALT>$<HASH>. A canonical
python script is included in share/rpcuser. The client then connects
normally using the rpcuser=<USERNAME>/rpcpassword=<PASSWORD> pair of
arguments. This option can be specified multiple times
.HP
\fB\-rpcport=\fR<port>
.IP
Listen for JSON\-RPC connections on <port> (default: 51473 or testnet:
51475)
.HP
\fB\-rpcallowip=\fR<ip>
.IP
Allow JSON\-RPC connections from specified source. Valid for <ip> are a
single IP (e.g. 1.2.3.4), a network/netmask (e.g. 1.2.3.4/255.255.255.0)
or a network/CIDR (e.g. 1.2.3.4/24). This option can be specified
multiple times
.HP
\fB\-rpcthreads=\fR<n>
.IP
Set the number of threads to service RPC calls (default: 4)
.HP
\fB\-blockspamfilter=\fR<n>
.IP
Use block spam filter (default: 1)
.HP
\fB\-blockspamfiltermaxsize=\fR<n>
.IP
Maximum size of the list of indexes in the block spam filter (default:
100)
.HP
\fB\-blockspamfiltermaxavg=\fR<n>
.IP
Maximum average size of an index occurrence in the block spam filter
(default: 10)
.SH COPYRIGHT
Copyright (C) 2009-2021 The Bitcoin Core Developers

Copyright (C) 2014-2021 The Dash Core Developers

Copyright (C) 2015-2021 The PIVX Core Developers

This is experimental software.

Distributed under the MIT software license, see the accompanying file COPYING
or <http://www.opensource.org/licenses/mit-license.php>.

This product includes UPnP software written by Thomas Bernard.


doc/release-notes.md
PIVX Core version *v5.5.0* is now available from:  <https://github.com/pivx-project/pivx/releases>

This is a new major version release, including various bug fixes and performance improvements.

Please report bugs using the issue tracker at github: <https://github.com/pivx-project/pivx/issues>

Mandatory Update
==============

PIVX Core v5.5.0 is a mandatory update for all users. This release contains new consensus rules and improvements that are not backwards compatible with older versions. Users will need to update their clients before enforcement of this update goes into effect.

Update enforcement is currently scheduled to go into effect at the following time:

```
Mainnet: block 3,715,200. ~Jan 20th 2023.
```
Masternodes will need to be restarted once both the masternode daemon and the controller wallet have been upgraded.
If you are running a Masternode over Tor, please read the "How To Upgrade" section.

Note: In preparation for the enforcement, upgraded peers will start rejecting non-upgraded peers few days before the enforcement block height, we recommend everyone to be updated at minimum a week before the final time.

Please report bugs using the issue tracker at github: <https://github.com/pivx-project/pivx/issues>


How to Upgrade
==============

If you are running an older version, shut it down. Wait until it has completely shut down (which might take a few minutes for older versions), then run the installer (on Windows) or just copy over /Applications/PIVX-Qt (on Mac) or pivxd/pivx-qt (on Linux).

Notable Changes
==============

New Rewards Structure
---------------------

Starting at block 3,715,200 The rewards structure is being adjusted after many months of discussion, analysis, and agreement from many in the community.

The new rewards are as follows:

|      Party     | Current Rewards | New Rewards |
| :---           |      :---:      |    :---:    |
| Staker         |        2        |      4      |
| Masternode     |        3        |      6      |
| Budget<sup>1</sup>        |        1        |      10     |

1: Budget coins are **ONLY** created if there is a valid passing budget proposal requesting a specific amount of coins to be created. This occurs roughly once every 30 days.

Network Activity Toggle
-----------------------

The GUI wallet now allows for the user to enable/disable all network activity from the Settings page. This can be handy if, for example, you need to restart your router or modem.

RPC Changes
-----------

### getblock additional verbosity

The `getblock` command now has an additional verbosity level. The command's second parameter has been changed from a boolean to and integer to support this.

- If verbosity is 0, returns a string that is serialized, hex-encoded data for block 'hash'.
- If verbosity is 1, returns an Object with information about block `hash`.
- If verbosity is 2, returns an Object with information about block `hash` and information about each transaction.

### getnewshieldaddress label support

The `getnewshieldaddress` command now supports an optional `label` parameter. If specified, the text label will be added to the wallet's address book.

### New setnetworkactive command

A new `setnetworkactive` command has been added to enable/disable all network activity. The command specs are as follows:

```
setnetworkactive true|false
Enable/Disable all p2p network activity.

Result:
status: (boolean) The final network activity status

Examples:
setnetworkactive true
setnetworkactive false
```

### New scantxoutset command

A new `scantxoutset` command has been added that allows for advanced searching of the unspent transaction outputs (UTXOs). As this is a highly advanced and complex command that won't apply to most end users, the full command spec won't be detailed here.

Please refer to the detailed spec by running `help scantxoutset` or refer to #2780.

*v5.5.0* Change log
==============

Detailed release notes follow. This is the complete list of PRs for this release version. For convenience in locating the code changes and accompanying discussion, both the pull request and git merge commit are mentioned.

### Consensus
- #2763 `f849beaf2c2` [Consensus] New block rewards (tohsnoom)
- #2770 `8d3f950bb40` [Consensus] Set testnet new rewards height (Fuzzbawls)
- #2772 `9be813c8c6d` bugfix new block reward payment and verification (PeterL73)
- #2781 `414be936487` [Consensus] Set v5.5 activation height for mainnet (Fuzzbawls)
- #2784 `3c63095b5fe` [Consensus] Bump v5.5 activation height for mainnet (Fuzzbawls)

### RPC
- #2732 `7216dc980f4` Remove cs_main lock from blockToJSON (random-zebra)
- #2734 `f79e009fe37` [RPC] Consistently use ParseHashV to validate hash inputs in rpc (random-zebra)
- #2746 `e3d54848845` Fix mnconnect missing parameter and rename cxxtimer.hpp (random-zebra)
- #2774 `768dfb5a5f1` [RPC] add include_delegated to vRPCConvertParams (PeterL73)
- #2779 `abb888b0ac7` [RPC] Backport getblock verbosity (tecnovert)
- #2780 `8cc21819d80` [RPC] Backport scantxoutset from BitCoin Core v0.17.2 (PeterL73)

### Build Systems
- #2650 `7e2d1aff1f8` Update bls-signatures subtree to latest version (furszy)
- #2694 `060a0a5656d` [Build] Bump master to 5.4.99 (furszy)
- #2695 `2d37006acb1` [CLI][Build] Guard libevent error messages (Fuzzbawls)
- #2702 `f2b45fded0c` [Core][Build] Fix atomic data races in BLS operations (Fuzzbawls)
- #2710 `5c5bb069aa9` [Build] Clean up additional coverage and test data (Fuzzbawls)
- #2731 `04ae0156cca` [Build] Add CMake option to compile with debug flags (Fuzzbawls)
- #2738 `37635f49076` [Build] Fix multi-line comment warning (Fuzzbawls)
- #2740 `0e757ad2c9b` [Lint] Introduce lint-includes.sh script (Fuzzbawls)
- #2759 `d2a30915ca2` [Build][CMake] Use CMake internal instead of custom command for moc (Fuzzbawls)
- #2767 `291ef05f538` [GA] Run unit/functional tests on native macOS 11 (Fuzzbawls)
- #2769 `bb8b073d108` [GA] Disable Unit tests for win64 build in GA (Fuzzbawls)

src/pivxd.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2014 The Bitcoin developers
// Copyright (c) 2014-2015 The Dash developers
// Copyright (c) 2015-2019 The PIVX developers
// Distributed under the MIT/X11 software license, see the accompanying
// file COPYING or https://www.opensource.org/licenses/mit-license.php.

#if defined(HAVE_CONFIG_H)
#include "config/pivx-config.h"
#endif

#include "chainparams.h"
#include "clientversion.h"
#include "fs.h"
#include "init.h"
#include "masternodeconfig.h"
#include "noui.h"
#include "shutdown.h"
#include "util/system.h"

#include <stdio.h>

/* Introduction text for doxygen: */

/*! \mainpage Developer documentation
 *
 * \section intro_sec Introduction
 *
 * This is the developer documentation of the reference client for an experimental new digital currency called PIVX (http://www.pivx.org),
 * which enables instant payments to anyone, anywhere in the world. PIVX uses peer-to-peer technology to operate
 * with no central authority: managing transactions and issuing money are carried out collectively by the network.
 *
 * The software is a community-driven open source project, released under the MIT license.
 *
 * \section Navigation
 * Use the buttons <code>Namespaces</code>, <code>Classes</code> or <code>Files</code> at the top of the page to start navigating the code.
 */

void WaitForShutdown()
{
    while (!ShutdownRequested()) {
        MilliSleep(200);
    }
    Interrupt();
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//
bool AppInit(int argc, char* argv[])
{
    bool fRet = false;

    //
    // Parameters
    //
    // If Qt is used, parameters/pivx.conf are parsed in qt/pivx.cpp's main()
    gArgs.ParseParameters(argc, argv);

    // Process help and version before taking care about datadir
    if (gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") || gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version")) {
        std::string strUsage = PACKAGE_NAME " Daemon version " + FormatFullVersion() + "\n";

        if (gArgs.IsArgSet("-version")) {
            strUsage += LicenseInfo();
        } else {
            strUsage += "\nUsage:  pivxd [options]                     Start " PACKAGE_NAME " Daemon\n";
            strUsage += "\n" + HelpMessage(HMM_BITCOIND);
        }

        fprintf(stdout, "%s", strUsage.c_str());
        return false;
    }

    try {
        if (!CheckDataDirOption()) {
            fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
            return false;
        }
        try {
            gArgs.ReadConfigFile(gArgs.GetArg("-conf", PIVX_CONF_FILENAME));
        } catch (const std::exception& e) {
            fprintf(stderr, "Error reading configuration file: %s\n", e.what());
            return false;
        }
        // Check for -testnet or -regtest parameter (Params() calls are only valid after this clause)
        try {
            SelectParams(gArgs.GetChainName());
        } catch(const std::exception& e) {
            fprintf(stderr, "Error: %s\n", e.what());
            return false;
        }

        // parse masternode.conf
        std::string strErr;
        if (!masternodeConfig.read(strErr)) {
            fprintf(stderr, "Error reading masternode configuration file: %s\n", strErr.c_str());
            return false;
        }

        // Error out when loose non-argument tokens are encountered on command line
        for (int i = 1; i < argc; i++) {
            if (!IsSwitchChar(argv[i][0])) {
                fprintf(stderr, "Error: Command line contains unexpected token '%s', see pivxd -h for a list of options.\n", argv[i]);
                return false;
            }
        }

        // -server defaults to true for pivxd but not for the GUI so do this here
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup()) {
            // UIError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitParameterInteraction()) {
            // UIError will have been called with detailed error, which ends up on console
            return false;
        }
        if (!AppInitSanityChecks()) {
            // UIError will have been called with detailed error, which ends up on console
            return false;
        }

#ifndef WIN32
        if (gArgs.GetBoolArg("-daemon", false)) {
            fprintf(stdout, "PIVX server starting\n");

            // Daemonize
            pid_t pid = fork();
            if (pid < 0) {
                fprintf(stderr, "Error: fork() returned %d errno %d\n", pid, errno);
                return false;
            }
            if (pid > 0) // Parent process, pid is child process id
            {
                return true;
            }
            // Child process falls through to rest of initialization

            pid_t sid = setsid();
            if (sid < 0)
                fprintf(stderr, "Error: setsid() returned %d errno %d\n", sid, errno);
        }
#endif

        // Set this early so that parameter interactions go to console
        fRet = AppInitMain();
    } catch (const std::exception& e) {
        PrintExceptionContinue(&e, "AppInit()");
    } catch (...) {
        PrintExceptionContinue(NULL, "AppInit()");
    }

    if (!fRet) {
        Interrupt();
    } else {
        WaitForShutdown();
    }
    Shutdown();

    return fRet;
}

int main(int argc, char* argv[])
{
#ifdef WIN32
    util::WinCmdLineArgs winArgs;
    std::tie(argc, argv) = winArgs.get();
#endif
    SetupEnvironment();

    // Connect pivxd signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? 0 : 1);
}
